lib/ide/src/controller/project.rs

```diff
@@ -78,47 +76,183 @@ impl Handle {
         let data = Controller::new_running(file_manager_transport);
         Self::new_from_data(data)
     }
+
+    /// Returns a text controller for given file path.
+    pub async fn open_text_file(&self, path:fmc::Path)
+    -> FallibleResult<controller::text::Handle> {
+        let cached = self.with_borrowed(|data| data.text_cache.get(&path));
+        match cached {
+            Some(controller) => Ok(controller),
+            None => {
+                let loaded = self.create_text_controller(path.clone()).await?;
+                //TODO[ao] Here we should make a better solution for case where we simultaneously
+                // load one module twice.
+                // This is duplicated with open_module, but it's not worth refactoring as its
+                // temporary solution.
+                let cached = self.with_borrowed(|data|
+                    match data.text_cache.entry(path) {
+                        Occupied(entry) => entry.get().clone_ref(),
+                        Vacant(entry)   => entry.insert(loaded)
+                    }
+                );
+                Ok(cached)
+            }
+        }
+    }
+
+    /// Returns a module controller which have module opened from file.
+    pub async fn open_module(&self, loc:ModuleLocation)
+    -> FallibleResult<controller::module::Handle> {
+        let cached = self.with_borrowed(|data| data.module_cache.get(&loc));
+        match cached {
+            Some(controller) => Ok(controller),
+            None => {
+                let loaded = self.create_module_controller(loc.clone()).await?;
+                //TODO[ao] Here we should make a better solution for case where we simultaneously
+                // load one module twice.
+                let cached = self.with_borrowed(|data|
+                    match data.module_cache.entry(loc) {
+                        Occupied(entry) => entry.get().clone_ref(),
+                        Vacant(entry)   => entry.insert(loaded)
+                    }
+                );
+                Ok(cached)
+            },
+        }
+    }
+
+    async fn create_text_controller(&self, path:fmc::Path)
+    -> FallibleResult<controller::text::Handle> {
+        match ModuleLocation::from_path(&path) {
+            Some(location) => {
+                let module = self.open_module(location).await?;
+                Ok(controller::text::Handle::new_for_module(module))
+            },
+            None => {
+                let fm = self.file_manager();
+                Ok(controller::text::Handle::new_for_plain_test(path,fm))
```

Comment:
This looks inconsistent to me. Why it "opens" a module (async operation) but creates a new text file?
I expect that there should be two methods for path: one that opens (i.e. awaits the download) target file (plain or module) and the other one that creates a new controller immediately (and whatever was in the file under path, if it existed, will be ignored and overwritten on save)

Reply:
But where it creates a new text file?

The main goal of this function is to provide TextController. TextController for module requires ModuleController, that's why we open it. For other files, it requires just file manager with path, so why we should open anything?

Reply:
Well, I renamed the "open_module" to "get_module_controller" and similarly "open_text_file" to "get_text_controller"